soh/soh/Enhancements/accessibility/accessibility.cpp

```diff
@@ -0,0 +1,727 @@
+            switch (cchar) {
+                case MESSAGE_NEWLINE:
+                    destBuf[destWriteIndex++] = (isListingChoices) ? '\n' : ' ';
+                    break;
+                case MESSAGE_THREE_CHOICE:
+                case MESSAGE_TWO_CHOICE:
+                    destBuf[destWriteIndex++] = '\n';
+                    isListingChoices = 1;
+                    break;
+                case MESSAGE_COLOR:
+                case MESSAGE_SHIFT:
+                case MESSAGE_TEXT_SPEED:
+                case MESSAGE_BOX_BREAK_DELAYED:
+                case MESSAGE_FADE:
+                case MESSAGE_ITEM_ICON:
+                    i++;
+                    break;
+                case MESSAGE_FADE2:
+                case MESSAGE_SFX:
+                case MESSAGE_TEXTID:
+                    i += 2;
+                    break;
+                default:
+                    break;
+            }
+        } else {
+            if (cchar <= 0x80) {
+                destBuf[destWriteIndex++] = cchar;
+            } else {
+                auto rchar = remap(cchar);
+                while (*rchar) {
+                    destBuf[destWriteIndex++] = *rchar++;
+                }
+            }
+        }
+    }
+    
+    destBuf[destWriteIndex] = 0;
+}
+
+void RegisterOnDialogMessageHook() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnDialogMessage>([]() {
+        if (!CVarGetInteger("gA11yTTS", 0)) return;
+        
+        MessageContext *msgCtx = &gPlayState->msgCtx;
+        
+        if (msgCtx->msgMode == MSGMODE_TEXT_NEXT_MSG || msgCtx->msgMode == MSGMODE_DISPLAY_SONG_PLAYED_TEXT_BEGIN || (msgCtx->msgMode == MSGMODE_TEXT_CONTINUING && msgCtx->stateTimer == 1)) {
+            ttsHasNewMessage = 1;
+        } else if (msgCtx->msgMode == MSGMODE_TEXT_DISPLAYING || msgCtx->msgMode == MSGMODE_TEXT_AWAIT_NEXT || msgCtx->msgMode == MSGMODE_TEXT_DONE || msgCtx->msgMode == MSGMODE_TEXT_DELAYED_BREAK
+                   || msgCtx->msgMode == MSGMODE_OCARINA_STARTING || msgCtx->msgMode == MSGMODE_OCARINA_PLAYING
+                   || msgCtx->msgMode == MSGMODE_DISPLAY_SONG_PLAYED_TEXT || msgCtx->msgMode == MSGMODE_DISPLAY_SONG_PLAYED_TEXT || msgCtx->msgMode == MSGMODE_SONG_PLAYED_ACT_BEGIN || msgCtx->msgMode == MSGMODE_SONG_PLAYED_ACT || msgCtx->msgMode == MSGMODE_SONG_PLAYBACK_STARTING || msgCtx->msgMode == MSGMODE_SONG_PLAYBACK || msgCtx->msgMode == MSGMODE_SONG_DEMONSTRATION_STARTING || msgCtx->msgMode == MSGMODE_SONG_DEMONSTRATION_SELECT_INSTRUMENT || msgCtx->msgMode == MSGMODE_SONG_DEMONSTRATION
+        ) {
+            if (ttsHasNewMessage) {
+                ttsHasMessage = 1;
+                ttsHasNewMessage = 0;
+                ttsCurrentHighlightedChoice = 0;
+                
+                uint16_t size = msgCtx->decodedTextLen;
+                Message_TTS_Decode(msgCtx->msgBufDecoded, ttsMessageBuf, 0, size);
+                SpeechSynthesizerSpeak(ttsMessageBuf, GetLanguageCode());
```

Comment:
Okay I ended up checking the `SpeechSynthesizerSpeak` implementation and it will pass the buffer address to another thread.
This might lead to undefined behavior here, if the speech synthesizer is still processing the buffer but another message writes into the `ttsMessageBuf`.